### PR TITLE
ability to trash recipes

### PR DESCRIPTION
**Pull request by mayanb**

### CODE REVIEW

File: ics/models.py [hunk 1]
```diff
@@ -657,7 +657,12 @@ class Alert(models.Model):
 class Recipe(models.Model):
 	product_type = models.ForeignKey(ProductType, related_name="recipes", on_delete=models.CASCADE)
 	process_type = models.ForeignKey(ProcessType, related_name="recipes", on_delete=models.CASCADE)
-	instructions = models.TextField()
+	instructions = models.TextField(null=True)
+	is_trashed = models.BooleanField(default=False)
```
Comment (sguichard) [inline ics/models.py]: Can we add an index to is_trashed since we'll be filtering by it?

File: ics/v9/views.py [hunk 2]
```diff
@@ -1055,3 +1056,22 @@ class TaskIngredientDetail(generics.RetrieveUpdateDestroyAPIView):
...
+  ingredients = json.loads((ingredients).decode("utf-8") )
+  for ingredient in ingredients:
+    product_type = ProductType.objects.get(pk=ingredient['product_type'])
+    process_type = ProcessType.objects.get(pk=ingredient['process_type'])
+    amount = ingredient['amount']
+    ing = Ingredient.objects.create(recipe=recipe, product_type=product_type, process_type=process_type, amount=amount)
+    created_ingredients.append(ing.id)
+  results = Ingredient.objects.filter(id__in=created_ingredients)
+  output_serializer = IngredientSerializer(results, many=True)
+  data = output_serializer.data[:]
```
Comment (sguichard) [inline ics/v9/views.py]: How come we need to copy this array? Maybe add a comment.

File: ics/v9/calculated_fields_serializers.py [hunk 3]
```diff
@@ -16,6 +16,14 @@ class NestedTaskSerializer(serializers.ModelSerializer):
 
 	task_ingredients = serializers.SerializerMethodField()
 	num_flagged_ancestors = serializers.IntegerField(read_only=True)
+	recipe_instructions = serializers.SerializerMethodField()
+
+	def get_recipe_instructions(self, task):
+		recipe = Recipe.objects.filter(product_type=task.product_type, process_type=task.process_type, is_trashed=False)
+		if recipe.count() > 0 and task.task_ingredients.count() > 0:
```
Comment (sguichard) [inline ics/v9/calculated_fields_serializers.py]: Is there a reason we wouldn't want to return the recipe instructions if the task doesn't have task_ingredients?

Comment (mayanb) [reply ics/v9/calculated_fields_serializers.py]: if the task doesn't have task_ingredients it means the task was created before the recipe was defined which means those instructions for that recipe shouldn't apply to that task. We also want to show recipes/instructions/ingredients for tasks that were created after a recipe is defined

Comment (sguichard) [reply ics/v9/calculated_fields_serializers.py]: Ok, it looks like this code will return the instructions for the non-deleted recipe, not the active recipe at the time the task is created.

I think it should be something like Recipe.objects.get(task.task_ingredients[0].ingredient.recipe.id) - but maybe more performant.

**Review by sguichard — APPROVED** [review]

**Pull request merged**